### PR TITLE
[hdpowerview] Return capabilities if capabilitiesOverride is not defined

### LINKED ISSUE #13029: [hdpowerview] Shades cannot be controlled through Hub v1 anymore
Shades can no longer be controlled when using Hub v1. Reported by @JeewanBahra.

## Expected Behavior
Like in 3.2, it should be possible to control shades from openHAB.

## Current Behavior
Controlling shades is broken in 3.3 in combination with Hub v1.

Initially logged:
```
2022-06-27 18:04:12.856 [TRACE] [rview.internal.HDPowerViewWebTargets] - API command GET http://192.168.3.226/api/fwversion/
2022-06-27 18:04:12.927 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON response = {"firmware":{"mainProcessor":{"name":"PowerView Hub","revision":1,"subRevision":1,"build":857}}}
2022-06-27 18:04:12.930 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Main processor firmware version received: PowerView Hub, 1.1.857
2022-06-27 18:04:12.933 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Polling for state
2022-06-27 18:04:12.935 [TRACE] [rview.internal.HDPowerViewWebTargets] - API command GET http://192.168.3.226/api/shades/
2022-06-27 18:04:13.027 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON response = {"shadeIds":[18210,64255,9142,28943],"shadeData":[{"id":18210,"name":"S2l0Y2hlbiBCbGluZA==","roomId":28574,"groupId":13505,"order":0,"type":6,"batteryStrength":177,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":64255,"name":"Q29uc2VydmF0b3J5IFJpZ2h0","roomId":28574,"groupId":13505,"order":1,"type":6,"batteryStrength":0,"batteryStatus":0, "positions":{"position1":65535,"posKind1":1}},{"id":9142,"name":"Q29uc2VydmF0b3J5IERvb3I=","roomId":28574,"groupId":13505,"order":2,"type":6,"batteryStrength":176,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":28943,"name":"Q29uc2VydmF0b3J5IExlZnQ=","roomId":28574,"groupId":13505,"order":3,"type":6,"batteryStrength":177,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}}]}
2022-06-27 18:04:13.036 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Received data for 4 shades
2022-06-27 18:04:13.039 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Updating shade '18210'
2022-06-27 18:04:13.042 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - Unable to set capabilities for shade 18210
2022-06-27 18:04:13.045 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - The 'shadeCapabilities' field has not yet been initialized
2022-06-27 18:04:13.051 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Updating shade '64255'
2022-06-27 18:04:13.054 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - Unable to set capabilities for shade 64255
2022-06-27 18:04:13.058 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - The 'shadeCapabilities' field has not yet been initialized
2022-06-27 18:04:13.063 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Updating shade '9142'
2022-06-27 18:04:13.066 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - Unable to set capabilities for shade 9142
2022-06-27 18:04:13.069 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - The 'shadeCapabilities' field has not yet been initialized
2022-06-27 18:04:13.074 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Updating shade '28943'
2022-06-27 18:04:13.076 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - Unable to set capabilities for shade 28943
2022-06-27 18:04:13.079 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - The 'shadeCapabilities' field has not yet been initialized
2022-06-27 18:04:13.084 [TRACE] [rview.internal.HDPowerViewWebTargets] - API command GET http://192.168.3.226/api/scenes/
2022-06-27 18:04:13.161 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON response = {"sceneIds":[],"sceneData":[]}
2022-06-27 18:04:13.163 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Received data for 0 scenes
2022-06-27 18:04:13.165 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Preserving scene channels, no changes detected
2022-06-27 18:04:13.167 [TRACE] [rview.internal.HDPowerViewWebTargets] - API command GET http://192.168.3.226/api/scenecollections/
2022-06-27 18:04:13.240 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON response = {"sceneCollectionIds": [],"sceneCollectionData":[]}
2022-06-27 18:04:13.242 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Received data for 0 sceneCollections
2022-06-27 18:04:13.245 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Preserving scene group channels, no changes detected
2022-06-27 18:04:13.247 [TRACE] [rview.internal.HDPowerViewWebTargets] - API command GET http://192.168.3.226/api/scheduledevents
2022-06-27 18:04:13.321 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON response = {"scheduledEventIds": [],"scheduledEventData":[]}
2022-06-27 18:04:13.324 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Received data for 0 scheduledEvents
2022-06-27 18:04:13.326 [DEBUG] [ternal.handler.HDPowerViewHubHandler] - Preserving automation channels, no changes detected
```

when sending a command (up/down):

```
2022-06-27 18:06:53.777 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON response = {"shade":{"id":28943,"name":"Q29uc2VydmF0b3J5IExlZnQ=","roomId":28574,"groupId":13505,"order":3,"type":6,"batteryStrength":177,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},"shadeData":[]}
2022-06-27 18:06:53.780 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - Unable to set capabilities for shade 28943
2022-06-27 18:06:53.784 [TRACE] [powerview.internal.api.ShadePosition] - setPosition(): capabilities=-- not in database -- (-1), coords=PRIMARY_POSITION, percent=100
2022-06-27 18:06:53.794 [TRACE] [rview.internal.HDPowerViewWebTargets] - API command PUT http://192.168.3.226/api/shades/28943
2022-06-27 18:06:53.797 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON command = {"shade":{"positions":{"posKind1":1,"position1":65535}}}
2022-06-27 18:06:53.876 [TRACE] [rview.internal.HDPowerViewWebTargets] - JSON response = {"shade":{"id":28943,"name":"Q29uc2VydmF0b3J5IExlZnQ=","roomId":28574,"groupId":13505,"order":3,"type":6,"batteryStrength":177,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}}}
2022-06-27 18:06:53.880 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - Unable to set capabilities for shade 28943
2022-06-27 18:06:53.882 [DEBUG] [rnal.handler.HDPowerViewShadeHandler] - The 'shadeCapabilities' field has not yet been initialized
```

Confirmed by @andrewfg to be a regression of #12902.
(initially was suspected a likely regression of #11698 which was introduced to fix #11557)

## Possible Solution
Make shade **type** and **capabilities** configurable, i.e.:

```
Thing shade shade1 [id=18210, type=6, capabilities=0]
```

Both should be optional. In order of precedence:
- If **type** is present, it will be used to look up **capabilities** in internal database.
- If **capabilities** is present, it will be used and will override **type** (if also present). If configured capabilities doesn't match capabilities for configured type, log this as a warning as configuration is inconsistent.
- If neither is present, default capabilities to 0. (or somehow reintroduce legacy mode similar to before #11698)
- When receiving **type** and **capabilities** from the API: When having no configuration, simply use values given by API. When having a configuration, check if configuration is consistent with received values. If not, provide warning **once** and use **configured** values.

## Steps to Reproduce
Try to send up/down commands to shades connected to a Hub v1.

## Your Environment
openHAB 3.3

### PR DESCRIPTION
**Pull request by andrewfg**
Fixes #13029 

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

### CODE REVIEW

**Comment by andrewfg** [comment]
The JAR is here..

[org.openhab.binding.hdpowerview-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/8998650/org.openhab.binding.hdpowerview-3.4.0-SNAPSHOT.jar.zip)

**Comment by jlaur** [comment]
I will review this later today.

File: bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java [hunk 1]
```diff
@@ -155,7 +155,7 @@ public int getCapabilities() {
          * @return 'typeCapabilities'.
          */
         public int getCapabilitiesOverride() {
-            return capabilitiesOverride;
+            return capabilitiesOverride < 0 ? capabilities : capabilitiesOverride;
```
Comment (jlaur) [inline bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java:158]: My understanding is that this method should simply provide the raw override value - the same way as `getCapabilities()` provides the raw capabilities (previous method). The fallback/override logic is in calling method, and probably should be kept and extended there?
https://github.com/openhab/openhab-addons/blob/0a2a85b448e9804fb88fc51e8f8ff22859cafe60/bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java#L335-L352

Also it would be nice if this is covered by a test.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for providing a quick fix. I have added one suggestion for consolidating the logic in one place, and another suggestion for making sure that it will work correctly no matter how it is implemented.

File: bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java [hunk 2]
```diff
@@ -335,16 +335,21 @@ public Capabilities getCapabilities(@Nullable Integer capabilitiesId) {
...
      * @param typeId the target shade type Id (to check if it has a 'capabilitiesOverride' value).
      * @param capabilitiesId the target capabilities value (when type Id does not have a 'capabilitiesOverride').
      * @return corresponding Capabilities class instance.
      */
     public Capabilities getCapabilities(int typeId, @Nullable Integer capabilitiesId) {
-        int targetCapabilities = TYPE_DATABASE.getOrDefault(typeId, new Type()).getCapabilitiesOverride();
+        Type type = TYPE_DATABASE.getOrDefault(typeId, new Type());
+        int targetCapabilities = type.getCapabilitiesOverride();
+        if (targetCapabilities < 0) {
+            targetCapabilities = type.getCapabilities();
```
Comment (jlaur) [inline bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java:351]: Is it intended that capabilities derived from type should take precedence over capabilities from response? It seems that order of precedence is now:
- Capabilities override from type.
- Capabilities from type.
- Capabilities from response.
- Default capabilities.

I would have expected:
- Capabilities override from type.
- Capabilities from response.
- Capabilities from type.
- Default capabilities.

Otherwise there are actually two levels of override from type, and capabilities from response is not actually used?

File: bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/ShadePositionTest.java [hunk 3]
```diff
@@ -403,4 +403,58 @@ public void testType44ShadePositionParsingShadeFullyDownVaneOpen() {
...
+        caps = db.getCapabilities(1, null);
+        assertEquals(0, caps.getValue());
+        /*
+         * - type 1 in database
+         * - inherent capabilities (0)
+         * - non matching external capabilities (1)
+         * => return inherent capabilities (0)
+         */
+        caps = db.getCapabilities(1, 1);
+        assertEquals(0, caps.getValue());
```
Comment (jlaur) [inline bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/ShadePositionTest.java:449]: Please see other comment. Here I would have expected 1 as directly provided by hub, i.e. taking precedence over the mismatching capabilities from type (0).

**Review by jlaur — CHANGES_REQUESTED** [review]
This is getting complex, so one more doubt.

File: bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java (hunk 2 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java:351]: > Is it intended that capabilities derived from type should take precedence over capabilities from response?

That is really a philosophical question. If the database is correct, then both approaches give the same result. We are also logging if the two capabilities do not match, so we can update the database. I don't have any objection to reversing the logic, so please let me know what do YOU prefer?

File: bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/ShadePositionTest.java (hunk 3 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/ShadePositionTest.java:449]: See sbove.

File: bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java (hunk 2 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java:351]: PS until now, except for the hub1 case, we have not seen any conflicts between the two approaches..

Comment (jlaur) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java:351]: IMHO capabilities from the response should be used when present, i.e. take precedence in case of mismatch between database content and actual capabilities. In a way it's a matter of trust - do we trust our database more than data provided by the hub. But also implementation-wise it seems confusing that type will **always override** capabilities (type->capabilities**Override** and then type->capabilities), so that capabilities from the hub is only used in the case where type is unknown.

Perhaps small difference... I'll not insist, although two overrides, where only one of them is called override by name, is kind of tickling my brain. It would also preserve existing behavior, as I assume PR was only intended to fix missing capabilities issue with v1 hub, not change precedence behavior in general?

Comment (andrewfg) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java:351]: Looking to the future, I suppose it is likely that HD introduces new types within the existing capabilities framework. But even in this case the scenario "type before capabilites" and the scenario "capabilities before type" will still produce the same result.

But let me sleep on this. I will get back to you tomorrow..

Comment (andrewfg) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/database/ShadeCapabilitiesDatabase.java:351]: Ok. I will do it as you suggest. I think also if no capabilities is found then the default value should not be -1 (unknown), but rather 0 (which means that at least some basic functionality will always exist..)

**Review by jlaur — APPROVED** [review]
Thanks for fixing the problem and also improving the code.

**Comment by jlaur** [comment]
@andrewfg - can you rephrase the PR title to be more user-oriented?

**Comment by andrewfg** [comment]
> rephrase the PR title to be more user-oriented

Done. And many thanks @jlaur for the review and merge. :)
